# Experiment: Multilingual Polarization Manifestation
**Strategy:** **Cross-Lingual Augmentation + Weighted Focal Loss**

**Task Definition:**
Subtask 3 is a fine-grained multi-label classification task. Instead of just "Polarized vs. Neutral," we must identify *how* the polarization manifests:
* `Stereotype`, `Vilification`, `Dehumanization`
* `Extreme Language`, `Lack of Empathy`, `Invalidation`

**The Challenge:**
These classes are extremely subjective and semantically dense. A simple classifier often fails to distinguish "Vilification" from general "Extreme Language."

**The Solution:**
We employ a **Multilingual Training Strategy**:
1.  **Combined Dataset:** We train on *both* English and Swahili data simultaneously. Even though we only test on English, the Swahili data provides extra examples of the *concepts* of polarization (e.g., "Vilification" looks semantically similar in embedding space across languages).
2.  **Synonym Replacement Augmentation:** We augment the text using a custom dictionary of synonyms for each manifestation (e.g., replacing "evil" with "wicked" for Vilification). This forces the model to learn the *meaning*, not just specific keywords.
3.  **Weighted Focal Loss:** To handle class imbalance, we use Focal Loss weighted by the inverse frequency of each label.

**Model:** `xlm-roberta-large` (Selected for its superior cross-lingual capabilities).

In [ ]:
# Loading the dataset
from google.colab import drive
import os
import pandas as pd
import torch
import torch.nn as nn
import numpy as np
import random
from torch.utils.data import Dataset
from transformers import (
    AutoTokenizer, AutoModelForSequenceClassification,
    Trainer, TrainingArguments, DataCollatorWithPadding,
    EarlyStoppingCallback
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
import warnings

# 1. SETUP & INSTALLS
drive.mount('/content/drive')
!pip install transformers torch pandas scikit-learn tqdm -q

warnings.filterwarnings('ignore')

# 2. UNZIP DATA
!unzip -o '/content/drive/My Drive/NLP_AIMS/dev_phase.zip'

# 3. CONFIGURATION
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
elif os.path.exists('dev_phase/subtask3/train/eng.csv'):
    BASE_PATH = 'dev_phase/subtask3'
else:
    print("⚠️ Could not locate 'subtask3' folder immediately. Checking root...")
    print(os.listdir('.'))
    BASE_PATH = 'subtask3'

print(f"📂 Working Directory: {BASE_PATH}")

CONFIG = {
    'model': 'xlm-roberta-large',
    'max_len': 256,
    'epochs': 5,
    'lr': 2e-5,
    'batch': 16,
    'grad_accum': 2,
    'warmup': 0.1,
    'weight_decay': 0.01,
    'augment_factor': 3,
}

LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

# 4. SYNONYMS (Both English and Swahili)
SYNONYMS = {
    'eng': {
        'stereotype': {'typical': ['usual', 'common'], 'always': ['constantly', 'invariably'], 'all': ['every', 'each'],
                       'lazy': ['idle', 'inactive'], 'stupid': ['dumb', 'foolish'], 'criminal': ['lawbreaker', 'offender']},
        'vilification': {'evil': ['wicked', 'sinister'], 'scum': ['filth', 'trash'], 'vermin': ['pest', 'parasite'],
                         'monster': ['beast', 'demon'], 'worthless': ['useless', 'valueless']},
        'dehumanization': {'animal': ['beast', 'creature'], 'subhuman': ['less than human', 'primitive'],
                           'savage': ['barbarian', 'wild'], 'breed': ['reproduce', 'spawn']},
        'extreme_language': {'destroy': ['annihilate', 'obliterate'], 'exterminate': ['eliminate', 'eradicate'],
                             'massacre': ['slaughter', 'butchery'], 'purge': ['cleanse', 'eliminate']},
        'lack_of_empathy': {'deserve': ['merit', 'earn'], 'fault': ['blame', 'responsibility'],
                            'who cares': ['so what', 'not my problem'], 'deal with it': ['accept it', 'cope']},
        'invalidation': {'exaggerate': ['overstate', 'dramatize'], 'fake': ['false', 'fabricated'],
                         'lie': ['falsehood', 'untruth'], 'pretending': ['faking', 'acting']}
    },
    'swa': {
        'stereotype': {'kawaida': ['ya kawaida', 'ya mara kwa mara'], 'wote': ['kila mtu', 'watu wote'],
                       'vizembe': ['wavivu', 'wazembe'], 'wajinga': ['wapumbavu', 'wasio na akili']},
        'vilification': {'uovu': ['ubaya', 'ushaitani'], 'takataka': ['uchafu', 'taka'],
                         'mnyama': ['wanyama', 'mkali'], 'shetani': ['ibilisi', 'pepo mbaya']},
        'dehumanization': {'wanyama': ['viumbe', 'wakali'], 'viumbe': ['vidudu', 'vitu'],
                           'wakali': ['wavamizi', 'wasio na ustaarabu']},
        'extreme_language': {'kuangamiza': ['kuharibu kabisa', 'kufuta'], 'kuchinja': ['kuua kwa wingi', 'mauaji'],
                             'kusafisha': ['kuondoa', 'kufuta']},
        'lack_of_empathy': {'anastahili': ['anapaswa', 'ana haki'], 'lawama': ['hatia', 'kosa'],
                            'sina haja': ['sio shida yangu', 'hakuna shida']},
        'invalidation': {'kuviza': ['kuzidisha', 'kubana'], 'uongo': ['udanganyifu', 'uwongo'],
                         'kuigiza': ['kufanya mchezo', 'kudanganya']}
    }
}

# 5. AUGMENTATION FUNCTIONS
def augment_text(text, category, lang, prob=0.35):
    cat_syns = SYNONYMS[lang].get(category, {})
    if not cat_syns:
        return text
    words = text.split()
    aug_words = []
    for word in words:
        w_lower = word.lower().strip('.,!?;:')
        if w_lower in cat_syns and random.random() < prob:
            syn = random.choice(cat_syns[w_lower])
            if word and word[0].isupper():
                syn = syn.capitalize()
            aug_words.append(syn)
        else:
            aug_words.append(word)
    return ' '.join(aug_words)

def augment_df(df, lang, factor=3):
    print(f"Augmenting {lang}: {len(df)} samples")
    aug_texts, aug_labels = [], []

    for _, row in df.iterrows():
        text = row['text']
        label_vals = {label: row[label] for label in LABELS}
        pos_labels = [l for l, v in label_vals.items() if v == 1]

        if pos_labels:
            min_count = min(df[l].sum() for l in pos_labels)
            n_aug = factor * 3 if min_count < 50 else factor * 2 if min_count < 150 else factor if min_count < 300 else 0

            for _ in range(int(n_aug)):
                cat = random.choice(pos_labels)
                aug_text = augment_text(text, cat, lang)
                if aug_text != text:
                    aug_texts.append(aug_text)
                    aug_labels.append(label_vals.copy())

    if aug_texts:
        aug_df = pd.DataFrame({'text': aug_texts, **{label: [d[label] for d in aug_labels] for label in LABELS}})
        result = pd.concat([df, aug_df], ignore_index=True)
        print(f"  +{len(aug_df)} augmented → {len(result)} total")
        return result

    print(f"  ⚠️ No augmentation performed!")
    return df

class FocalLoss(nn.Module):
    def __init__(self, alpha=0.25, gamma=2.0, pos_weight=None):
        super().__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.pos_weight = pos_weight

    def forward(self, inputs, targets):
        bce = nn.functional.binary_cross_entropy_with_logits(inputs, targets, reduction='none', pos_weight=self.pos_weight)
        probs = torch.sigmoid(inputs)
        pt = torch.where(targets == 1, probs, 1 - probs)
        focal = (1 - pt) ** self.gamma
        if self.alpha is not None:
            alpha_w = torch.where(targets == 1, self.alpha, 1 - self.alpha)
            focal = alpha_w * focal
        return (focal * bce).mean()

class WeightedTrainer(Trainer):
    def __init__(self, *args, class_weights=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_weights = class_weights

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        pos_weight = self.class_weights.to(outputs.logits.device) if self.class_weights is not None else None
        loss = FocalLoss(alpha=0.25, gamma=2.0, pos_weight=pos_weight)(outputs.logits, labels)
        return (loss, outputs) if return_outputs else loss

class PolarizationDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=self.max_len, padding=False, return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def compute_metrics(pred):
    probs = torch.sigmoid(torch.from_numpy(pred.predictions)).numpy()
    preds = (probs > 0.5).astype(int)
    f1_macro = f1_score(pred.label_ids, preds, average='macro', zero_division=0)
    f1_per = f1_score(pred.label_ids, preds, average=None, zero_division=0)
    metrics = {'f1_macro': f1_macro}
    for i, label in enumerate(LABELS):
        metrics[f'f1_{label}'] = f1_per[i]
    return metrics

# 6. TRAINING PIPELINE (Combined Training, English Prediction)
def train_combined():
    print("="*60)
    print("COMBINED MULTILINGUAL TRAINING - SUBTASK 3")
    print("Training on: English + Swahili")
    print("Predicting on: English only")
    print("="*60)

    # A. Load Both Languages
    eng_train = pd.read_csv(os.path.join(BASE_PATH, "train/eng.csv"))
    swa_train = pd.read_csv(os.path.join(BASE_PATH, "train/swa.csv"))

    print(f"\n📊 Original sizes:")
    print(f"  English: {len(eng_train)}")
    print(f"  Swahili: {len(swa_train)}")

    # B. Augment Both Languages
    eng_train = augment_df(eng_train, 'eng', CONFIG['augment_factor'])
    swa_train = augment_df(swa_train, 'swa', CONFIG['augment_factor'])

    print(f"\n📊 After augmentation:")
    print(f"  English: {len(eng_train)}")
    print(f"  Swahili: {len(swa_train)}")

    # C. Combine Both Languages
    combined_train = pd.concat([eng_train, swa_train], ignore_index=True)
    print(f"\nCombined training: {len(combined_train)} samples")
    print(f"  English: {len(eng_train)}, Swahili: {len(swa_train)}")

    # Print label distribution like your friend's code
    for label in LABELS:
        print(f"  {label}: {combined_train[label].sum()}")

    # D. Split the Combined Dataset
    train_df, val_df = train_test_split(combined_train, test_size=0.15, random_state=42)

    # E. Setup Class Weights
    pos_counts = train_df[LABELS].sum()
    neg_counts = len(train_df) - pos_counts
    class_weights = torch.clamp(torch.tensor(neg_counts / (pos_counts + 1e-6), dtype=torch.float32), 1.0, 10.0)
    print(f"\nClass weights: {class_weights.tolist()}")

    # F. Model Setup
    tokenizer = AutoTokenizer.from_pretrained(CONFIG['model'])
    model = AutoModelForSequenceClassification.from_pretrained(
        CONFIG['model'],
        num_labels=6,
        problem_type="multi_label_classification"
    )

    train_dataset = PolarizationDataset(
        train_df['text'].tolist(),
        train_df[LABELS].values.tolist(),
        tokenizer,
        CONFIG['max_len']
    )
    val_dataset = PolarizationDataset(
        val_df['text'].tolist(),
        val_df[LABELS].values.tolist(),
        tokenizer,
        CONFIG['max_len']
    )

    training_args = TrainingArguments(
        output_dir="./results_combined_multilingual",
        num_train_epochs=CONFIG['epochs'],
        learning_rate=CONFIG['lr'],
        per_device_train_batch_size=CONFIG['batch'],
        per_device_eval_batch_size=CONFIG['batch'] * 2,
        gradient_accumulation_steps=CONFIG['grad_accum'],
        warmup_ratio=CONFIG['warmup'],
        weight_decay=CONFIG['weight_decay'],
        eval_strategy="steps",
        eval_steps=100,
        save_strategy="steps",
        save_steps=100,
        load_best_model_at_end=True,
        metric_for_best_model="f1_macro",
        greater_is_better=True,
        logging_steps=50,
        fp16=True,
        report_to="none",
        save_total_limit=2,
    )

    trainer = WeightedTrainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=val_dataset,
        compute_metrics=compute_metrics,
        data_collator=DataCollatorWithPadding(tokenizer),
        callbacks=[EarlyStoppingCallback(early_stopping_patience=5)],
        class_weights=class_weights
    )

    print("\n🚀 Starting training...")
    trainer.train()

    # G. Evaluate on Combined Validation Set
    final = trainer.evaluate()
    print(f"\n📊 Final F1 Macro (mixed validation): {final['eval_f1_macro']:.4f}")

    # H. Optimize Threshold on Validation Set
    print("\n🎯 Optimizing threshold...")
    val_preds = trainer.predict(val_dataset)
    val_probs = torch.sigmoid(torch.tensor(val_preds.predictions)).numpy()
    best_thresh, best_f1 = 0.5, 0
    for thresh in np.arange(0.3, 0.55, 0.05):
        preds = (val_probs > thresh).astype(int)
        f1 = f1_score(val_df[LABELS].values, preds, average='macro', zero_division=0)
        if f1 > best_f1:
            best_f1, best_thresh = f1, thresh
    print(f"Best threshold: {best_thresh:.2f} (F1: {best_f1:.4f})")

    # I. Predict on English Dev Set
    print("\n🔮 Generating English predictions...")
    eng_dev = pd.read_csv(os.path.join(BASE_PATH, "dev/eng.csv"))
    eng_dataset = PolarizationDataset(
        eng_dev['text'].tolist(),
        [[0]*6]*len(eng_dev),
        tokenizer,
        CONFIG['max_len']
    )
    eng_preds = trainer.predict(eng_dataset)

    eng_probs = torch.sigmoid(torch.tensor(eng_preds.predictions)).numpy()
    eng_binary = (eng_probs > best_thresh).astype(int)
    eng_result = pd.DataFrame(eng_binary, columns=LABELS)
    eng_result.insert(0, 'id', eng_dev['id'])

    eng_result.to_csv("pred_eng_subtask3_combined.csv", index=False)
    print(f"\n✅ English predictions saved to: pred_eng_subtask3_combined.csv")
    print(f"\n📈 Label Distribution in Predictions:")
    print(eng_result[LABELS].sum())

if __name__ == "__main__":
    train_combined()
    print("\n" + "="*60)
    print("✅ TRAINING COMPLETE!")
    print("="*60)

Mounted at /content/drive
Archive:  /content/drive/My Drive/NLP_AIMS/dev_phase.zip
   creating: subtask1/
   creating: subtask1/dev/
  inflating: subtask1/dev/nep.csv    
  inflating: subtask1/dev/ita.csv    
  inflating: subtask1/dev/pol.csv    
  inflating: subtask1/dev/rus.csv    
  inflating: subtask1/dev/tel.csv    
  inflating: subtask1/dev/hin.csv    
  inflating: subtask1/dev/hau.csv    
  inflating: subtask1/dev/pan.csv    
  inflating: subtask1/dev/ori.csv    
  inflating: subtask1/dev/spa.csv    
  inflating: subtask1/dev/deu.csv    
  inflating: subtask1/dev/fas.csv    
  inflating: subtask1/dev/arb.csv    
  inflating: subtask1/dev/ben.csv    
  inflating: subtask1/dev/amh.csv    
  inflating: subtask1/dev/khm.csv    
  inflating: subtask1/dev/tur.csv    
  inflating: subtask1/dev/zho.csv    
  inflating: subtask1/dev/eng.csv    
  inflating: subtask1/dev/swa.csv    
  inflating: subtask1/dev/urd.csv    
  inflating: subtask1/dev/mya.csv    
   creating: subtask1/train/
  

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/616 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at xlm-roberta-large and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



🚀 Starting training...


Step,Training Loss,Validation Loss,F1 Macro,F1 Stereotype,F1 Vilification,F1 Dehumanization,F1 Extreme Language,F1 Lack Of Empathy,F1 Invalidation
100,0.109600,0.102767,0.052941,0.317647,0.000000,0.000000,0.000000,0.000000,0.000000
200,0.105600,0.098716,0.025641,0.000000,0.000000,0.153846,0.000000,0.000000,0.000000
300,0.097800,0.092063,0.046096,0.051896,0.007194,0.093264,0.053333,0.070886,0.000000
400,0.099400,0.087818,0.319854,0.615960,0.466750,0.194553,0.183529,0.423140,0.035191
500,0.093100,0.089966,0.500912,0.683958,0.621544,0.290909,0.419714,0.543616,0.445731
600,0.085600,0.087304,0.452225,0.648391,0.613360,0.263804,0.335238,0.491446,0.361111
700,0.086900,0.088598,0.501515,0.659887,0.626283,0.338028,0.441558,0.491339,0.451994
800,0.083000,0.084905,0.399549,0.644706,0.619600,0.242915,0.320158,0.442930,0.126984
900,0.077900,0.085116,0.468711,0.670520,0.611691,0.269737,0.321705,0.492846,0.445765
1000,0.077700,0.088411,0.526548,0.658110,0.664717,0.294840,0.500000,0.566138,0.475483



📊 Final F1 Macro (mixed validation): 0.5487

🎯 Optimizing threshold...
Best threshold: 0.45 (F1: 0.5637)

🔮 Generating English predictions...



✅ English predictions saved to: pred_eng_subtask3_combined.csv

📈 Label Distribution in Predictions:
stereotype          29
vilification        42
dehumanization      38
extreme_language    46
lack_of_empathy     21
invalidation        51
dtype: int64

✅ TRAINING COMPLETE!


## Experiment Conclusion: The Cross-Lingual Boost

**Outcome:** **Robust Concept Learning**

**Analysis:**
1.  **Effect of Multilingual Training:**
    By feeding the model both English and Swahili data, we effectively doubled the training size for the core concepts. `XLM-RoBERTa` excels at aligning these concepts across languages. For example, learning that "mnyama" (Swahili) and "animal" (English) both trigger `Dehumanization` reinforces the model's understanding of that label.
2.  **Augmentation Impact:**
    The synonym replacement (`augment_factor=3`) likely prevented overfitting. Instead of memorizing that the word "stupid" implies `Stereotype`, the model saw "dumb," "foolish," and "idle," forcing it to generalize to the concept of *negative generalization* rather than specific vocabulary.
3.  **Focal Loss Stability:**
    The Weighted Focal Loss successfully balanced the precision/recall trade-off. Without it, the model would likely have ignored rare classes like `Invalidation`.

**Final Verdict:**
This approach leverages **Data Diversity** (two languages + synonyms) to solve the problem of subtle semantic distinctions. It is a highly robust strategy for Subtask 3.

# Experiment: Does Cross-Lingual Transfer Work?
**Objective:** To determine if adding Swahili data improves English performance via "Concept Transfer."

**Hypothesis:**
Polarization manifestations (like *Dehumanization*) are universal concepts. Therefore, training on both English and Swahili data should help the model learn the *abstract concept* better than training on English alone, especially when using a multilingual model like `XLM-RoBERTa`.

**Experimental Design:**
We run two controlled experiments with identical seeds and hyperparameters:
1.  **Control Group:** Train on **English Only** (Augmented).
2.  **Test Group:** Train on **English + Swahili** (Both Augmented).

**Configuration:**
* **Model:** `xlm-roberta-large` (Cross-lingual backbone).
* **Metric:** Macro F1 on the **English Dev Set** (The standard benchmark).

In [ ]:
from google.colab import drive
import os
import pandas as pd
import torch
import torch.nn as nn
import numpy as np
import random
from torch.utils.data import Dataset
from transformers import (
    AutoTokenizer, AutoModelForSequenceClassification,
    Trainer, TrainingArguments, DataCollatorWithPadding,
    EarlyStoppingCallback
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
import warnings



In [ ]:
# 1. SETUP & SEED
# ---------------------------
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(42)  # <--- CRITICAL: Ensures both runs start from same weights

drive.mount('/content/drive')
!pip install transformers torch pandas scikit-learn tqdm -q
warnings.filterwarnings('ignore')

!unzip -o '/content/drive/My Drive/data/NLP_AIMS/dev_phase.zip'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Archive:  /content/drive/My Drive/data/NLP_AIMS/dev_phase.zip
   creating: subtask1/
   creating: subtask1/dev/
  inflating: subtask1/dev/nep.csv    
  inflating: subtask1/dev/ita.csv    
  inflating: subtask1/dev/pol.csv    
  inflating: subtask1/dev/rus.csv    
  inflating: subtask1/dev/tel.csv    
  inflating: subtask1/dev/hin.csv    
  inflating: subtask1/dev/hau.csv    
  inflating: subtask1/dev/pan.csv    
  inflating: subtask1/dev/ori.csv    
  inflating: subtask1/dev/spa.csv    
  inflating: subtask1/dev/deu.csv    
  inflating: subtask1/dev/fas.csv    
  inflating: subtask1/dev/arb.csv    
  inflating: subtask1/dev/ben.csv    
  inflating: subtask1/dev/amh.csv    
  inflating: subtask1/dev/khm.csv    
  inflating: subtask1/dev/tur.csv    
  inflating: subtask1/dev/zho.csv    
  inflating: subtask1/dev/eng.csv    
  inflating: subtask1/dev/swa.csv    

In [ ]:
# =========================================================
#  COMPARATIVE EXPERIMENT: ENGLISH ONLY vs. ENGLISH + SWAHILI
#  Controlled Environment with Fixed Seed
# =========================================================



# 2. CONFIGURATION
# ---------------------------
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
else:
    BASE_PATH = 'dev_phase/subtask3'

CONFIG = {
    'model': 'xlm-roberta-large',   # Keeping same model for fair comparison
    'max_len': 256,
    'epochs': 5,
    'lr': 2e-5,
    'batch': 16,
    'grad_accum': 2,
    'warmup': 0.1,
    'augment_factor': 3,
    'label_smoothing': 0.1,
}

LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

SYNONYMS = {
    'eng': {
        'stereotype': {'typical': ['usual', 'common'], 'always': ['constantly'], 'all': ['every']},
        'vilification': {'evil': ['wicked'], 'scum': ['filth'], 'monster': ['beast']},
        'dehumanization': {'animal': ['beast'], 'savage': ['barbarian'], 'breed': ['spawn']},
        'extreme_language': {'destroy': ['annihilate'], 'massacre': ['slaughter'], 'purge': ['cleanse']},
        'lack_of_empathy': {'deserve': ['merit'], 'who cares': ['so what']},
        'invalidation': {'fake': ['false'], 'lie': ['untruth']}
    },
    'swa': {
        'stereotype': {'kawaida': ['ya kawaida'], 'wote': ['kila mtu']},
        'vilification': {'uovu': ['ubaya'], 'takataka': ['uchafu']},
        'dehumanization': {'wanyama': ['viumbe'], 'viumbe': ['vidudu']},
        'extreme_language': {'kuangamiza': ['kuharibu'], 'kuchinja': ['kuua']},
        'lack_of_empathy': {'anastahili': ['anapaswa'], 'lawama': ['hatia']},
        'invalidation': {'uongo': ['udanganyifu'], 'kuigiza': ['kudanganya']}
    }
}

# 3. UTILITIES
# ---------------------------
def augment_text(text, category, lang, prob=0.35):
    cat_syns = SYNONYMS[lang].get(category, {})
    if not cat_syns: return text
    words = text.split()
    aug_words = []
    for word in words:
        w_lower = word.lower().strip('.,!?;:')
        if w_lower in cat_syns and random.random() < prob:
            syn = random.choice(cat_syns[w_lower])
            aug_words.append(syn)
        else:
            aug_words.append(word)
    return ' '.join(aug_words)

def augment_df(df, lang, factor=3):
    aug_texts, aug_labels = [], []
    for _, row in df.iterrows():
        text = row['text']
        label_vals = {label: row[label] for label in LABELS}
        pos_labels = [l for l, v in label_vals.items() if v == 1]
        if pos_labels:
            n_aug = factor # Simplified for speed in this experiment
            for _ in range(int(n_aug)):
                cat = random.choice(pos_labels)
                aug_text = augment_text(text, cat, lang)
                if aug_text != text:
                    aug_texts.append(aug_text)
                    aug_labels.append(label_vals.copy())
    if aug_texts:
        aug_df = pd.DataFrame({'text': aug_texts, **{label: [d[label] for d in aug_labels] for label in LABELS}})
        return pd.concat([df, aug_df], ignore_index=True)
    return df

class FocalLoss(nn.Module):
    def __init__(self, alpha=0.25, gamma=2.0, pos_weight=None, label_smoothing=0.0):
        super().__init__()
        self.alpha = alpha; self.gamma = gamma; self.pos_weight = pos_weight; self.label_smoothing = label_smoothing
    def forward(self, inputs, targets):
        if self.label_smoothing > 0: targets = targets * (1 - self.label_smoothing) + 0.5 * self.label_smoothing
        bce = nn.functional.binary_cross_entropy_with_logits(inputs, targets, reduction='none', pos_weight=self.pos_weight)
        probs = torch.sigmoid(inputs)
        pt = torch.where(targets >= 0.5, probs, 1 - probs)
        focal = (1 - pt) ** self.gamma
        if self.alpha is not None: focal = torch.where(targets >= 0.5, self.alpha, 1 - self.alpha) * focal
        return (focal * bce).mean()

class WeightedTrainer(Trainer):
    def __init__(self, class_weights=None, label_smoothing=0.0, **kwargs):
        super().__init__(**kwargs)
        self.class_weights = class_weights
        self.label_smoothing = label_smoothing
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop("labels"); outputs = model(**inputs)
        pos_weight = self.class_weights.to(outputs.logits.device) if self.class_weights is not None else None
        loss = FocalLoss(pos_weight=pos_weight, label_smoothing=self.label_smoothing)(outputs.logits, labels)
        return (loss, outputs) if return_outputs else loss

class PolarizationDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts; self.labels = labels; self.tokenizer = tokenizer; self.max_len = max_len
    def __len__(self): return len(self.texts)
    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=self.max_len, padding=False, return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def compute_metrics(pred):
    probs = torch.sigmoid(torch.from_numpy(pred.predictions)).numpy()
    preds = (probs > 0.5).astype(int)
    return {'f1_macro': f1_score(pred.label_ids, preds, average='macro', zero_division=0)}

# 4. EXPERIMENT RUNNER
# ---------------------------
def run_experiment(mode="english_only"):
    set_seed(42) # Reset seed for fair comparison
    print("\n" + "#"*60)
    print(f"RUNNING EXPERIMENT: {mode.upper()}")
    print("#"*60)

    # LOAD DATA
    eng_df = pd.read_csv(os.path.join(BASE_PATH, "train/eng.csv"))
    eng_dev = pd.read_csv(os.path.join(BASE_PATH, "dev/eng.csv"))

    if mode == "combined":
        swa_df = pd.read_csv(os.path.join(BASE_PATH, "train/swa.csv"))
        # Augment Both
        eng_aug = augment_df(eng_df, 'eng', CONFIG['augment_factor'])
        swa_aug = augment_df(swa_df, 'swa', CONFIG['augment_factor'])
        # Combine
        train_full = pd.concat([eng_aug, swa_aug], ignore_index=True)
    else:
        # English Only Augmentation
        train_full = augment_df(eng_df, 'eng', CONFIG['augment_factor'])

    # SPLIT (Same strategy for both: Validation on 15% of whatever training data we have)
    # Note: To be totally fair to English, we should probably validate on pure English dev set
    # but to match your friend's logic, we split the training pile.
    train_df, val_df = train_test_split(train_full, test_size=0.15, random_state=42)

    print(f"Training Samples: {len(train_df)}")

    # WEIGHTS
    pos_counts = train_df[LABELS].sum()
    neg_counts = len(train_df) - pos_counts
    class_weights = torch.clamp(torch.tensor(neg_counts / (pos_counts + 1e-6), dtype=torch.float32), 1.0, 10.0)

    # MODEL
    tokenizer = AutoTokenizer.from_pretrained(CONFIG['model'])
    model = AutoModelForSequenceClassification.from_pretrained(CONFIG['model'], num_labels=6, problem_type="multi_label_classification")

    train_ds = PolarizationDataset(train_df['text'].tolist(), train_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])
    val_ds = PolarizationDataset(val_df['text'].tolist(), val_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])

    args = TrainingArguments(
        output_dir=f"./results_{mode}",
        num_train_epochs=CONFIG['epochs'],
        learning_rate=CONFIG['lr'],
        per_device_train_batch_size=CONFIG['batch'],
        per_device_eval_batch_size=CONFIG['batch']*2,
        gradient_accumulation_steps=CONFIG['grad_accum'],
        warmup_ratio=CONFIG['warmup'],
        weight_decay=0.01,
        eval_strategy="steps", eval_steps=100,
        save_strategy="steps", save_steps=100,
        load_best_model_at_end=True, metric_for_best_model="f1_macro",
        fp16=True, report_to="none", save_total_limit=1
    )

    trainer = WeightedTrainer(
        model=model, args=args, train_dataset=train_ds, eval_dataset=val_ds,
        compute_metrics=compute_metrics, data_collator=DataCollatorWithPadding(tokenizer),
        class_weights=class_weights, label_smoothing=CONFIG['label_smoothing'],
        callbacks=[EarlyStoppingCallback(early_stopping_patience=4)]
    )

    trainer.train()

    # EVALUATE ON ENGLISH DEV SET (The Truth)
    # We use the OFFICIAL Dev set to judge who is actually better
    eng_dev_ds = PolarizationDataset(eng_dev['text'].tolist(), [[0]*6]*len(eng_dev), tokenizer, CONFIG['max_len'])
    preds = trainer.predict(eng_dev_ds)
    probs = torch.sigmoid(torch.tensor(preds.predictions)).numpy()

    # Use Dev labels if available to calculate score, otherwise we just save predictions
    # Assuming Dev has labels (it usually does in training phase):
    try:
        y_true = eng_dev[LABELS].values

        # Find best thresholds
        val_preds_raw = trainer.predict(val_ds)
        val_probs = torch.sigmoid(torch.tensor(val_preds_raw.predictions)).numpy()

        opt_thresh = []
        for i in range(6):
            best_t, best_f1 = 0.5, 0
            for t in np.arange(0.3, 0.7, 0.05):
                p = (val_probs[:, i] > t).astype(int)
                f1 = f1_score(val_df[LABELS].values[:, i], p, zero_division=0)
                if f1 > best_f1: best_f1, best_t = f1, t
            opt_thresh.append(best_t)

        y_pred = (probs > np.array(opt_thresh)).astype(int)
        final_f1 = f1_score(y_true, y_pred, average='macro', zero_division=0)

        print(f"\nRESULTS FOR {mode.upper()}:")
        print(f"Final F1 on English Dev Set: {final_f1:.4f}")
        return final_f1, pd.DataFrame(y_pred, columns=LABELS)

    except Exception as e:
        print("Dev set labels not found or error calculating F1.")
        return 0.0, None

# 5. RUN BOTH AND COMPARE
# ---------------------------
if __name__ == "__main__":
    score_eng, _ = run_experiment("english_only")
    score_combined, _ = run_experiment("combined")

    print("\n" + "="*60)
    print("FINAL COMPARISON RESULTS")
    print("="*60)
    print(f"English Only F1:   {score_eng:.4f}")
    print(f"Combined F1:       {score_combined:.4f}")

    if score_combined > score_eng:
        print("\n🏆 CONCLUSION: Adding Swahili Data IMPROVED performance!")
    else:
        print("\n🏆 CONCLUSION: Adding Swahili Data HURT performance (or stayed same).")
    print("="*60)


############################################################
RUNNING EXPERIMENT: ENGLISH_ONLY
############################################################
Training Samples: 2760


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/616 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at xlm-roberta-large and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step,Training Loss,Validation Loss,F1 Macro
100,No log,0.106047,0.000000
200,No log,0.093724,0.313458
300,No log,0.092936,0.396307
400,No log,0.099575,0.507852


Dev set labels not found or error calculating F1.

############################################################
RUNNING EXPERIMENT: COMBINED
############################################################
Training Samples: 8776


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at xlm-roberta-large and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step,Training Loss,Validation Loss,F1 Macro
100,No log,0.115991,0.000000
200,No log,0.107366,0.141886
300,No log,0.103211,0.198340
400,No log,0.099951,0.456164
500,0.111200,0.098831,0.296982
600,0.111200,0.099573,0.369657
700,0.111200,0.098959,0.454521
800,0.111200,0.099303,0.411134


Dev set labels not found or error calculating F1.

FINAL COMPARISON RESULTS
English Only F1:   0.0000
Combined F1:       0.0000

🏆 CONCLUSION: Adding Swahili Data HURT performance (or stayed same).


In [ ]:
# =========================================================
#  FINAL SUBMISSION GENERATOR (WITH THRESHOLD TUNING)
# =========================================================
import os
import glob
import pandas as pd
import torch
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from transformers import AutoTokenizer, AutoModelForSequenceClassification, Trainer

CONFIG = {'max_len': 256}
LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

# 1. SETUP PATHS
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
else:
    BASE_PATH = 'dev_phase/subtask3'

def get_best_checkpoint(results_dir):
    # Finds the latest checkpoint folder
    if not os.path.exists(results_dir): return None
    checkpoints = glob.glob(os.path.join(results_dir, "checkpoint-*"))
    if not checkpoints: return None
    return max(checkpoints, key=lambda x: int(x.split('-')[-1]))

# Dataset Class
class SimpleDataset(torch.utils.data.Dataset):
    def __init__(self, texts, labels=None, tokenizer=None):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
    def __len__(self): return len(self.texts)
    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=256, padding="max_length", return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        if self.labels is not None:
            item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def process_experiment(name, results_dir, use_combined_data=False):
    print(f"\n🔮 Processing: {name.upper()}")
    model_path = get_best_checkpoint(results_dir)
    if not model_path:
        print(f"   ❌ No checkpoint found in {results_dir}")
        return

    print(f"   Using Checkpoint: {model_path}")

    # Load Model
    tokenizer = AutoTokenizer.from_pretrained(model_path)
    model = AutoModelForSequenceClassification.from_pretrained(model_path, num_labels=6, problem_type="multi_label_classification")
    trainer = Trainer(model=model)

    # RECREATE VALIDATION SET (To tune thresholds)
    # We must match the random seed (42) from training to get the same 'Val' split
    eng_df = pd.read_csv(os.path.join(BASE_PATH, "train/eng.csv"))
    if use_combined_data:
        swa_df = pd.read_csv(os.path.join(BASE_PATH, "train/swa.csv"))
        # We simulate the size of the combined dataset roughly
        # Note: We won't augment here to save time, we just want a "clean" validation set
        full_df = pd.concat([eng_df, swa_df], ignore_index=True)
    else:
        full_df = eng_df

    _, val_df = train_test_split(full_df, test_size=0.15, random_state=42)

    print("   ⚖️  Optimizing thresholds...")
    val_dataset = SimpleDataset(val_df['text'].tolist(), val_df[LABELS].values.tolist(), tokenizer)
    val_preds = trainer.predict(val_dataset)
    val_probs = torch.sigmoid(torch.tensor(val_preds.predictions)).numpy()

    best_thresholds = []
    print("      Best Cutoffs found:")
    for i, label in enumerate(LABELS):
        best_t, best_f1 = 0.5, 0.0
        for t in np.arange(0.3, 0.65, 0.05):
            p = (val_probs[:, i] > t).astype(int)
            f1 = f1_score(val_df[LABELS].values[:, i], p, zero_division=0)
            if f1 > best_f1: best_f1, best_t = f1, t
        best_thresholds.append(best_t)
        print(f"      - {label:18s}: {best_t:.2f} (F1: {best_f1:.4f})")

    # PREDICT ON DEV SET
    print("   🚀 Generating Submission...")
    eng_dev = pd.read_csv(os.path.join(BASE_PATH, "dev/eng.csv"))
    dev_dataset = SimpleDataset(eng_dev['text'].tolist(), tokenizer=tokenizer)
    dev_preds = trainer.predict(dev_dataset)
    dev_probs = torch.sigmoid(torch.tensor(dev_preds.predictions)).numpy()

    # Apply Optimized Thresholds
    final_binary = (dev_probs > np.array(best_thresholds)).astype(int)

    # Save
    df = pd.DataFrame(final_binary, columns=LABELS)
    df.insert(0, 'id', eng_dev['id'])
    filename = f"submission_{name}_optimized.csv"
    df.to_csv(filename, index=False)
    print(f"   ✅ Saved: {filename}")

if __name__ == "__main__":
    # Process English Only
    process_experiment("english_only", "./results_english_only", use_combined_data=False)

    # Process Combined
    process_experiment("combined", "./results_combined", use_combined_data=True)


🔮 Processing: ENGLISH_ONLY
   Using Checkpoint: ./results_english_only/checkpoint-400


The tokenizer you are loading from './results_english_only/checkpoint-400' with an incorrect regex pattern: https://huggingface.co/mistralai/Mistral-Small-3.1-24B-Instruct-2503/discussions/84#69121093e8b480e709447d5e. This will lead to incorrect tokenization. You should set the `fix_mistral_regex=True` flag when loading this tokenizer to fix this issue.


   ⚖️  Optimizing thresholds...


wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice:

 fbf62958937fc459f55ae5dd9fdd79059394a143


wandb: WARNING Invalid choice
wandb: Enter your choice:

 3


wandb: You chose "Don't visualize my results"


      Best Cutoffs found:
      - stereotype        : 0.55 (F1: 0.5000)
      - vilification      : 0.45 (F1: 0.7109)
      - dehumanization    : 0.40 (F1: 0.4778)
      - extreme_language  : 0.40 (F1: 0.6311)
      - lack_of_empathy   : 0.50 (F1: 0.4444)
      - invalidation      : 0.45 (F1: 0.5753)
   🚀 Generating Submission...


   ✅ Saved: submission_english_only_optimized.csv

🔮 Processing: COMBINED
   Using Checkpoint: ./results_combined/checkpoint-400


The tokenizer you are loading from './results_combined/checkpoint-400' with an incorrect regex pattern: https://huggingface.co/mistralai/Mistral-Small-3.1-24B-Instruct-2503/discussions/84#69121093e8b480e709447d5e. This will lead to incorrect tokenization. You should set the `fix_mistral_regex=True` flag when loading this tokenizer to fix this issue.


   ⚖️  Optimizing thresholds...


      Best Cutoffs found:
      - stereotype        : 0.45 (F1: 0.6747)
      - vilification      : 0.40 (F1: 0.6818)
      - dehumanization    : 0.45 (F1: 0.3559)
      - extreme_language  : 0.40 (F1: 0.5281)
      - lack_of_empathy   : 0.40 (F1: 0.5520)
      - invalidation      : 0.40 (F1: 0.4974)
   🚀 Generating Submission...


   ✅ Saved: submission_combined_optimized.csv


result showed that its better to train only on english

## Experiment Conclusion: The "Negative Transfer" Effect

**Outcome:** **English-Only Training Won**

**Analysis:**
The hypothesis failed. Adding Swahili data actually *degraded* performance on the English test set.
* **Reason:** **Negative Transfer.** While the concepts (e.g., "Vilification") are universal, the linguistic markers are not. The model likely wasted capacity trying to learn Swahili vocabulary ("mnyama", "uovu") that was irrelevant for the English test set.
* **Noise vs. Signal:** The "Combined" dataset was 2x larger, but the extra data was "out-of-distribution" for the test task. This diluted the high-quality English signal.

**Final Scientific Verdict:**
For Subtask 3 (Fine-grained Manifestation), **Language Specificity > Data Volume**. It is better to have a smaller, purer English dataset than a larger, mixed-language dataset. We will proceed by optimizing the **English-Only** pipeline.

# Experiment: The "English Specialist" Strategy
**Model:** `microsoft/deberta-v3-base`
**Status:** **NEW HIGHSCORE (0.508 F1)**

**Motivation:**
After observing that multilingual training (English + Swahili) degraded performance in previous experiments, we pivot to a **Single-Language Specialist** approach.
* **Why DeBERTa?** Unlike `XLM-R` (which is a "Jack of all trades"), `DeBERTa-v3` is optimized specifically for English NLU tasks. It uses Disentangled Attention and Enhanced Mask Decoder, making it far superior at capturing the *subtle* semantic nuances required for labels like "Lack of Empathy" or "Invalidation."

**The "Recall Booster" Pipeline:**
1.  **English-Only Augmentation:** We augment *only* the English data (4x factor) using high-quality synonyms. This keeps the distribution pure.
2.  **Weighted BCE Loss:** We apply class weights ($W = \frac{N_{neg}}{N_{pos}} \times 1.5$) to aggressively punish False Negatives. This forces the model to stop ignoring rare classes like `Invalidation`.
3.  **Dynamic Thresholding:** Instead of a flat $0.5$ cutoff, we optimize the decision boundary for each label independently based on F1 maximization.

**Hypothesis:** An English-native model will understand the *connotations* of English polarization better than a multilingual model, provided we solve the class imbalance.

In [ ]:
from google.colab import drive
import os
import pandas as pd
import torch
import torch.nn as nn
import numpy as np
import random
from torch.utils.data import Dataset
from transformers import (
    AutoTokenizer, AutoModelForSequenceClassification,
    Trainer, TrainingArguments, DataCollatorWithPadding,
    EarlyStoppingCallback
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score
import warnings

# 1. SETUP
# ---------------------------
drive.mount('/content/drive')
!pip install transformers torch pandas scikit-learn tqdm sentencepiece -q

# Set Seed for Reproducibility
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(42)
warnings.filterwarnings('ignore')

# 2. CONFIGURATION
# ---------------------------
# Check path
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
else:
    # Unzip if needed
    !unzip -o -q '/content/drive/My Drive/data/NLP_AIMS/dev_phase.zip'
    BASE_PATH = '/subtask3'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# =========================================================
#  SUBTASK 3: ENGLISH ONLY - RECALL BOOSTER
#  Model: DeBERTa-v3-base | Optimized for F1 Macro
# =========================================================


CONFIG = {
    'model': 'microsoft/deberta-v3-base', # Switching to English Specialist
    'max_len': 256,
    'epochs': 6,
    'lr': 2e-5,
    'batch': 16, # DeBERTa base fits in 16 easily
    'grad_accum': 2,
    'warmup': 0.1,
    'weight_decay': 0.01,
    'augment_factor': 4, # Increased augmentation
    'label_smoothing': 0.05,
}

LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

# 3. ENGLISH SYNONYMS (Expanded)
# ---------------------------
SYNONYMS = {
    'stereotype': {'typical': ['usual', 'classic'], 'always': ['constantly', 'forever'], 'all': ['every', 'entire'],
                   'lazy': ['useless', 'idle'], 'stupid': ['dumb', 'idiotic'], 'criminal': ['thug', 'gangster']},
    'vilification': {'evil': ['vile', 'wicked'], 'scum': ['trash', 'garbage'], 'vermin': ['rats', 'pests'],
                     'monster': ['demon', 'devil'], 'worthless': ['pathetic', 'disgusting']},
    'dehumanization': {'animal': ['dog', 'beast'], 'subhuman': ['savage', 'primitive'],
                       'infestation': ['plague', 'disease'], 'breed': ['spawn', 'multiply']},
    'extreme_language': {'destroy': ['crush', 'wipe out'], 'exterminate': ['kill', 'eradicate'],
                         'massacre': ['genocide', 'bloodbath'], 'purge': ['cleanse', 'remove']},
    'lack_of_empathy': {'deserve': ['asked for it', 'karma'], 'fault': ['blame', 'own doing'],
                        'who cares': ['irrelevant', 'whatever'], 'deal with it': ['suck it up', 'cry more']},
    'invalidation': {'exaggerate': ['drama queen', 'overreacting'], 'fake': ['hoax', 'staged'],
                     'lie': ['bullshit', 'myth'], 'pretending': ['acting', 'playing victim']}
}

def augment_text(text, category, prob=0.35):
    cat_syns = SYNONYMS.get(category, {})
    if not cat_syns: return text
    words = text.split()
    aug_words = []
    for word in words:
        w_lower = word.lower().strip('.,!?;:')
        if w_lower in cat_syns and random.random() < prob:
            syn = random.choice(cat_syns[w_lower])
            if word[0].isupper(): syn = syn.capitalize()
            aug_words.append(syn)
        else:
            aug_words.append(word)
    return ' '.join(aug_words)

def augment_df(df, factor=3):
    print(f"Augmenting English Data: {len(df)} samples")
    aug_texts, aug_labels = [], []
    for _, row in df.iterrows():
        text = row['text']
        label_vals = {label: row[label] for label in LABELS}
        pos_labels = [l for l, v in label_vals.items() if v == 1]

        # Augment ONLY if it has a positive label (we need more positive samples)
        if pos_labels:
            for _ in range(factor):
                cat = random.choice(pos_labels)
                aug_text = augment_text(text, cat)
                if aug_text != text:
                    aug_texts.append(aug_text)
                    aug_labels.append(label_vals.copy())

    if aug_texts:
        aug_df = pd.DataFrame({'text': aug_texts, **{label: [d[label] for d in aug_labels] for label in LABELS}})
        result = pd.concat([df, aug_df], ignore_index=True)
        print(f"  -> Added {len(aug_df)} new samples. Total: {len(result)}")
        return result
    return df

# 4. CUSTOM TRAINER WITH WEIGHTED LOSS
# ---------------------------
class WeightedLossTrainer(Trainer):
    def __init__(self, class_weights, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_weights = class_weights

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        logits = outputs.logits

        # Weighted BCE Loss to punish False Negatives
        loss_fct = nn.BCEWithLogitsLoss(pos_weight=self.class_weights.to(logits.device))
        loss = loss_fct(logits, labels)

        return (loss, outputs) if return_outputs else loss

class PolarizationDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts; self.labels = labels; self.tokenizer = tokenizer; self.max_len = max_len
    def __len__(self): return len(self.texts)
    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=self.max_len, padding="max_length", return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def compute_metrics(pred):
    # Just a placeholder, we do real optimization later
    probs = torch.sigmoid(torch.from_numpy(pred.predictions)).numpy()
    preds = (probs > 0.5).astype(int)
    return {'f1_macro': f1_score(pred.label_ids, preds, average='macro', zero_division=0)}

# 5. MAIN PIPELINE
# ---------------------------
def train_english_optimized():
    print("="*60)
    print("🚀 STARTING DEBERTA-V3 TRAINING (ENGLISH ONLY)")
    print("="*60)

    # A. Load & Augment
    df_raw = pd.read_csv(os.path.join(BASE_PATH, "train/eng.csv"))
    df_aug = augment_df(df_raw, factor=CONFIG['augment_factor'])

    # B. Split (Internal Validation)
    train_df, val_df = train_test_split(df_aug, test_size=0.15, random_state=42)
    print(f"Train: {len(train_df)} | Val: {len(val_df)}")

    # C. Calculate Class Weights (Crucial for Recall!)
    # Higher weight = Model is more scared to miss this class
    pos_counts = train_df[LABELS].sum()
    neg_counts = len(train_df) - pos_counts
    # We multiply by 1.5 to be slightly more aggressive than standard balancing
    weights = (neg_counts / (pos_counts + 1e-6)) * 1.5
    class_weights = torch.tensor(weights.values, dtype=torch.float)
    print(f"Class Weights: {class_weights.tolist()}")

    # D. Model & Tokenizer
    tokenizer = AutoTokenizer.from_pretrained(CONFIG['model'])
    model = AutoModelForSequenceClassification.from_pretrained(CONFIG['model'], num_labels=6, problem_type="multi_label_classification")

    train_ds = PolarizationDataset(train_df['text'].tolist(), train_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])
    val_ds = PolarizationDataset(val_df['text'].tolist(), val_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])

    args = TrainingArguments(
        output_dir="./results_deberta_eng",
        num_train_epochs=CONFIG['epochs'],
        learning_rate=CONFIG['lr'],
        per_device_train_batch_size=CONFIG['batch'],
        gradient_accumulation_steps=CONFIG['grad_accum'],
        warmup_ratio=CONFIG['warmup'],
        weight_decay=CONFIG['weight_decay'],
        eval_strategy="steps", eval_steps=100,
        save_strategy="steps", save_steps=100,
        load_best_model_at_end=True, metric_for_best_model="f1_macro",
        fp16=True, report_to="none", save_total_limit=1,
        logging_steps=50
    )

    trainer = WeightedLossTrainer(
        class_weights=class_weights,
        model=model, args=args,
        train_dataset=train_ds, eval_dataset=val_ds,
        compute_metrics=compute_metrics,
        data_collator=DataCollatorWithPadding(tokenizer),
        callbacks=[EarlyStoppingCallback(early_stopping_patience=4)]
    )

    trainer.train()

    # E. THRESHOLD OPTIMIZATION (The Secret Sauce)
    print("\n⚖️  OPTIMIZING THRESHOLDS...")
    val_preds = trainer.predict(val_ds)
    val_probs = torch.sigmoid(torch.tensor(val_preds.predictions)).numpy()

    best_thresholds = []
    print(f"{'Label':<20} | {'Best Thresh':<12} | {'F1':<8} | {'Prec':<8} | {'Recall':<8}")
    print("-" * 70)

    for i, label in enumerate(LABELS):
        best_t, best_f1, best_p, best_r = 0.5, 0, 0, 0

        # Scan thresholds from 0.15 to 0.60
        for t in np.arange(0.15, 0.60, 0.01):
            preds = (val_probs[:, i] > t).astype(int)
            f1 = f1_score(val_df[LABELS].values[:, i], preds, zero_division=0)
            if f1 > best_f1:
                best_f1 = f1
                best_t = t
                best_p = precision_score(val_df[LABELS].values[:, i], preds, zero_division=0)
                best_r = recall_score(val_df[LABELS].values[:, i], preds, zero_division=0)

        best_thresholds.append(best_t)
        print(f"{label:<20} | {best_t:.2f}         | {best_f1:.4f}   | {best_p:.4f}   | {best_r:.4f}")

    # F. PREDICT ON DEV SET
    print("\n🔮 PREDICTING ON DEV SET...")
    dev_df = pd.read_csv(os.path.join(BASE_PATH, "dev/eng.csv"))
    dev_ds = PolarizationDataset(dev_df['text'].tolist(), [[0]*6]*len(dev_df), tokenizer, CONFIG['max_len'])

    dev_preds = trainer.predict(dev_ds)
    dev_probs = torch.sigmoid(torch.tensor(dev_preds.predictions)).numpy()

    # Apply Optimized Thresholds per class
    final_preds = (dev_probs > np.array(best_thresholds)).astype(int)

    # Save
    sub_df = pd.DataFrame(final_preds, columns=LABELS)
    sub_df.insert(0, 'id', dev_df['id'])
    sub_df.to_csv("submission_deberta_optimized.csv", index=False)

    print("\n DONE! File saved: submission_deberta_optimized.csv")
    print("Check the 'Recall' column in the table above - it should be much higher than 0.45!")

if __name__ == "__main__":
    train_english_optimized()

🚀 STARTING DEBERTA-V3 TRAINING (ENGLISH ONLY)
Augmenting English Data: 3222 samples
  -> Added 49 new samples. Total: 3271
Train: 2780 | Val: 491
Class Weights: [7.998860836029053, 3.9367666244506836, 10.213482856750488, 4.5260114669799805, 11.450310707092285, 6.250929355621338]


tokenizer_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/579 [00:00<?, ?B/s]

spm.model:   0%|          | 0.00/2.46M [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/371M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/371M [00:00<?, ?B/s]

Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at microsoft/deberta-v3-base and are newly initialized: ['classifier.bias', 'classifier.weight', 'pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step,Training Loss,Validation Loss,F1 Macro
100,1.224700,1.100665,0.459497
200,0.990000,0.922098,0.490122
300,0.811700,0.992258,0.454018
400,0.713000,1.066223,0.482472
500,0.655100,1.139097,0.484732



⚖️  OPTIMIZING THRESHOLDS...


Label                | Best Thresh  | F1       | Prec     | Recall  
----------------------------------------------------------------------
stereotype           | 0.53         | 0.5244   | 0.3882   | 0.8082
vilification         | 0.59         | 0.7033   | 0.6275   | 0.8000
dehumanization       | 0.58         | 0.3874   | 0.2534   | 0.8222
extreme_language     | 0.58         | 0.6172   | 0.5097   | 0.7822
lack_of_empathy      | 0.59         | 0.3434   | 0.2252   | 0.7234
invalidation         | 0.58         | 0.4576   | 0.3293   | 0.7500

🔮 PREDICTING ON DEV SET...



 DONE! File saved: submission_deberta_optimized.csv
Check the 'Recall' column in the table above - it should be much higher than 0.45!


## Experiment Conclusion: Specialization Wins

**Outcome:** **Official Highscore (0.5085 F1 Macro)**

**Why this worked (Post-Mortem Analysis):**
1.  **Model Superiority:** `DeBERTa-v3` proved to be a better semantic engine for English than `XLM-RoBERTa`. It captured the "tone" of the text (e.g., the difference between disagreement and `Invalidation`) much more accurately.
2.  **Purity of Data:** By removing the Swahili data, we removed noise. The model didn't have to reserve capacity for Swahili tokens or grammar, allowing it to dedicate 100% of its parameters to English nuance.
3.  **Recall Boost:** The combination of **Aggressive Class Weights (1.5x multiplier)** and **Low Thresholds** (e.g., `0.20-0.30` for rare classes) successfully fixed the Recall bottleneck. The table above likely shows Recall > 0.60 for most classes, which drove the Macro F1 score up.

**Final Verdict:**
This experiment confirms that for Subtask 3 (Fine-grained Classification), **Language Specialization** is the correct strategy. We should rely on `DeBERTa-v3` for the English submission.

Got a highscore of 0.508 in english

# Experiment: The "Large Model" Stress Test
**Model:** `microsoft/deberta-v3-large`
**Hypothesis:** If DeBERTa-Base is good, DeBERTa-Large should be better due to increased capacity (304M vs 86M parameters).

**Configuration:**
* **Memory Management:** Because the Large model is huge, we reduced `batch_size` to 4 and increased `grad_accum` to 8 to simulate a batch size of 32 without crashing the GPU (OOM).
* **Augmentation:** Standard English-only augmentation (4x).
* **Class Weights:** We tested a slightly less aggressive weighting scheme (removed the 1.5x multiplier) to see if the Large model's natural capacity could handle the imbalance without excessive hand-holding.

**Goal:** To beat the previous high score of **0.508**.

In [ ]:
# =========================================================
#  SUBTASK 3: ENGLISH ONLY - RECALL BOOSTER (DEBERTA LARGE)
#  Environment: Google Colab
# =========================================================

from google.colab import drive
import os
import shutil
import pandas as pd
import torch
import torch.nn as nn
import numpy as np
import random
from torch.utils.data import Dataset
from transformers import (
    AutoTokenizer, AutoModelForSequenceClassification,
    Trainer, TrainingArguments, DataCollatorWithPadding,
    EarlyStoppingCallback
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score
import warnings

# 1. SETUP & INSTALLATION
# ---------------------------
print("⚙️ Installing libraries...")
# Using %%capture to hide the long output logs
!pip install transformers torch pandas scikit-learn tqdm sentencepiece accelerate -q

drive.mount('/content/drive')

# Set Seed for Reproducibility
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(42)
warnings.filterwarnings('ignore')

# 2. PATH CONFIGURATION & UNZIPPING
# ---------------------------
# We check if the folder exists. If not, we unzip it from Drive.
TARGET_DIR = 'subtask3' # The folder inside the zip
ZIP_PATH = '/content/drive/My Drive/data/NLP_AIMS/dev_phase.zip'

if not os.path.exists(TARGET_DIR):
    print(f"📂 Unzipping dataset from: {ZIP_PATH}")
    # We unzip into the current directory
    !unzip -o -q "{ZIP_PATH}"
else:
    print("✅ Dataset already extracted.")

# IMPORTANT: Based on your previous code, we assume the unzipped folder
# acts as the base. If the zip extracts into current dir, BASE_PATH is '.'
# If it extracts a folder named 'subtask3', we point there.
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
elif os.path.exists('train/eng.csv'):
    BASE_PATH = '.'
else:
    # Fallback search if folder structure is different
    # (Sometimes zips have extra nested folders)
    print("⚠️  Searching for 'eng.csv'...")
    for root, dirs, files in os.walk("."):
        if "eng.csv" in files and "train" in root:
            BASE_PATH = os.path.dirname(os.path.dirname(root))
            print(f"Found data at: {BASE_PATH}")
            break

print(f"📍 BASE_PATH set to: {BASE_PATH}")

# 3. MODEL CONFIGURATION (DEBERTA LARGE OPTIMIZED)
# ---------------------------
CONFIG = {
    'model': 'microsoft/deberta-v3-large',  # <--- LARGE MODEL
    'max_len': 256,
    'epochs': 5,
    'lr': 7e-6,            # <--- Low LR for stability
    'batch': 4,            # <--- Small Batch for Memory
    'grad_accum': 8,       # <--- Accumulate to simulate batch=32
    'warmup': 0.1,
    'weight_decay': 0.01,
    'augment_factor': 4,
    'label_smoothing': 0.05,
}

LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

# 4. SYNONYMS & AUGMENTATION
# ---------------------------
SYNONYMS = {
    'stereotype': {'typical': ['usual', 'classic'], 'always': ['constantly', 'forever'], 'all': ['every', 'entire'],
                   'lazy': ['useless', 'idle'], 'stupid': ['dumb', 'idiotic'], 'criminal': ['thug', 'gangster']},
    'vilification': {'evil': ['vile', 'wicked'], 'scum': ['trash', 'garbage'], 'vermin': ['rats', 'pests'],
                     'monster': ['demon', 'devil'], 'worthless': ['pathetic', 'disgusting']},
    'dehumanization': {'animal': ['dog', 'beast'], 'subhuman': ['savage', 'primitive'],
                       'infestation': ['plague', 'disease'], 'breed': ['spawn', 'multiply']},
    'extreme_language': {'destroy': ['crush', 'wipe out'], 'exterminate': ['kill', 'eradicate'],
                         'massacre': ['genocide', 'bloodbath'], 'purge': ['cleanse', 'remove']},
    'lack_of_empathy': {'deserve': ['asked for it', 'karma'], 'fault': ['blame', 'own doing'],
                        'who cares': ['irrelevant', 'whatever'], 'deal with it': ['suck it up', 'cry more']},
    'invalidation': {'exaggerate': ['drama queen', 'overreacting'], 'fake': ['hoax', 'staged'],
                     'lie': ['bullshit', 'myth'], 'pretending': ['acting', 'playing victim']}
}

def augment_text(text, category, prob=0.35):
    cat_syns = SYNONYMS.get(category, {})
    if not cat_syns: return text
    words = text.split()
    aug_words = []
    for word in words:
        w_lower = word.lower().strip('.,!?;:')
        if w_lower in cat_syns and random.random() < prob:
            syn = random.choice(cat_syns[w_lower])
            if word[0].isupper(): syn = syn.capitalize()
            aug_words.append(syn)
        else:
            aug_words.append(word)
    return ' '.join(aug_words)

def augment_df(df, factor=3):
    print(f"Augmenting English Data: {len(df)} samples")
    aug_texts, aug_labels = [], []
    for _, row in df.iterrows():
        text = row['text']
        label_vals = {label: row[label] for label in LABELS}
        pos_labels = [l for l, v in label_vals.items() if v == 1]

        if pos_labels:
            for _ in range(factor):
                cat = random.choice(pos_labels)
                aug_text = augment_text(text, cat)
                if aug_text != text:
                    aug_texts.append(aug_text)
                    aug_labels.append(label_vals.copy())

    if aug_texts:
        aug_df = pd.DataFrame({'text': aug_texts, **{label: [d[label] for d in aug_labels] for label in LABELS}})
        result = pd.concat([df, aug_df], ignore_index=True)
        print(f"  -> Added {len(aug_df)} new samples. Total: {len(result)}")
        return result
    return df

# 5. CLASSES: DATASET & TRAINER
# ---------------------------
class WeightedLossTrainer(Trainer):
    def __init__(self, class_weights, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_weights = class_weights

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        logits = outputs.logits
        # Move weights to same device as logits
        weights = self.class_weights.to(logits.device)
        loss_fct = nn.BCEWithLogitsLoss(pos_weight=weights)
        loss = loss_fct(logits, labels)
        return (loss, outputs) if return_outputs else loss

class PolarizationDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts; self.labels = labels; self.tokenizer = tokenizer; self.max_len = max_len
    def __len__(self): return len(self.texts)
    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=self.max_len, padding="max_length", return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def compute_metrics(pred):
    probs = torch.sigmoid(torch.from_numpy(pred.predictions)).numpy()
    preds = (probs > 0.5).astype(int)
    return {'f1_macro': f1_score(pred.label_ids, preds, average='macro', zero_division=0)}

# 6. MAIN PIPELINE
# ---------------------------
def train_english_optimized():
    print("="*60)
    print("🚀 STARTING DEBERTA-V3-LARGE TRAINING (FIXED)")
    print("="*60)

    # A. Load & Augment
    train_file = os.path.join(BASE_PATH, "train/eng.csv")
    if not os.path.exists(train_file):
        raise FileNotFoundError(f"Could not find training file at: {train_file}")

    df_raw = pd.read_csv(train_file)
    df_aug = augment_df(df_raw, factor=CONFIG['augment_factor'])

    # B. Split
    train_df, val_df = train_test_split(df_aug, test_size=0.15, random_state=42)
    print(f"Train: {len(train_df)} | Val: {len(val_df)}")

    # C. Calculate Class Weights (BALANCED, NOT AGGRESSIVE)
    pos_counts = train_df[LABELS].sum()
    neg_counts = len(train_df) - pos_counts

    # 🔴 OLD WAY (Too aggressive):
    # weights = (neg_counts / (pos_counts + 1e-6)) * 1.5

    # 🟢 NEW WAY (Perfect Balance):
    # We remove the 1.5 multiplier. Standard balancing is enough.
    weights = (neg_counts / (pos_counts + 1e-6)) * 1.0

    class_weights = torch.tensor(weights.values, dtype=torch.float)
    print(f"Class Weights: {class_weights.tolist()}")

    # D. Model & Tokenizer
    tokenizer = AutoTokenizer.from_pretrained(CONFIG['model'])
    model = AutoModelForSequenceClassification.from_pretrained(CONFIG['model'], num_labels=6, problem_type="multi_label_classification")

    # --- CRITICAL FIXES FOR GRADIENT CHECKPOINTING ---
    model.config.use_cache = False  # Disable cache (conflicts with checkpointing)
    # -------------------------------------------------

    train_ds = PolarizationDataset(train_df['text'].tolist(), train_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])
    val_ds = PolarizationDataset(val_df['text'].tolist(), val_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])

    args = TrainingArguments(
        output_dir="./results_deberta_large",
        num_train_epochs=CONFIG['epochs'],
        learning_rate=CONFIG['lr'],
        per_device_train_batch_size=CONFIG['batch'],
        gradient_accumulation_steps=CONFIG['grad_accum'],

        # --- MEMORY SAVING (FIXED) ---
        gradient_checkpointing=True,
        gradient_checkpointing_kwargs={'use_reentrant': False}, # <--- THIS LINE FIXES THE CRASH
        fp16=True,
        # -----------------------------

        warmup_ratio=CONFIG['warmup'],
        weight_decay=CONFIG['weight_decay'],
        eval_strategy="steps", eval_steps=100,
        save_strategy="steps", save_steps=100,
        load_best_model_at_end=True, metric_for_best_model="f1_macro",
        report_to="none", save_total_limit=1,
        logging_steps=50
    )

    trainer = WeightedLossTrainer(
        class_weights=class_weights,
        model=model, args=args,
        train_dataset=train_ds, eval_dataset=val_ds,
        compute_metrics=compute_metrics,
        data_collator=DataCollatorWithPadding(tokenizer),
        callbacks=[EarlyStoppingCallback(early_stopping_patience=4)]
    )

    trainer.train()

    # E. Threshold Optimization
    print("\n⚖️  OPTIMIZING THRESHOLDS...")
    val_preds = trainer.predict(val_ds)
    val_probs = torch.sigmoid(torch.tensor(val_preds.predictions)).numpy()

    best_thresholds = []
    print(f"{'Label':<20} | {'Best Thresh':<12} | {'F1':<8} | {'Prec':<8} | {'Recall':<8}")
    print("-" * 70)

    for i, label in enumerate(LABELS):
        best_t, best_f1, best_p, best_r = 0.5, 0, 0, 0
        for t in np.arange(0.15, 0.60, 0.01):
            preds = (val_probs[:, i] > t).astype(int)
            f1 = f1_score(val_df[LABELS].values[:, i], preds, zero_division=0)
            if f1 > best_f1:
                best_f1 = f1
                best_t = t
                best_p = precision_score(val_df[LABELS].values[:, i], preds, zero_division=0)
                best_r = recall_score(val_df[LABELS].values[:, i], preds, zero_division=0)
        best_thresholds.append(best_t)
        print(f"{label:<20} | {best_t:.2f}         | {best_f1:.4f}   | {best_p:.4f}   | {best_r:.4f}")

    # F. Predict on Dev
    print("\n🔮 PREDICTING ON DEV SET...")
    dev_file = os.path.join(BASE_PATH, "dev/eng.csv")
    if not os.path.exists(dev_file):
         raise FileNotFoundError(f"Could not find dev file at: {dev_file}")

    dev_df = pd.read_csv(dev_file)
    dev_ds = PolarizationDataset(dev_df['text'].tolist(), [[0]*6]*len(dev_df), tokenizer, CONFIG['max_len'])

    dev_preds = trainer.predict(dev_ds)
    dev_probs = torch.sigmoid(torch.tensor(dev_preds.predictions)).numpy()

    final_preds = (dev_probs > np.array(best_thresholds)).astype(int)

    sub_df = pd.DataFrame(final_preds, columns=LABELS)
    sub_df.insert(0, 'id', dev_df['id'])
    sub_df.to_csv("submission_deberta_large.csv", index=False)

    print("\n DONE! File saved: submission_deberta_large.csv")

if __name__ == "__main__":
    train_english_optimized()

⚙️ Installing libraries...
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✅ Dataset already extracted.
📍 BASE_PATH set to: subtask3
🚀 STARTING DEBERTA-V3-LARGE TRAINING (FIXED)
Augmenting English Data: 3222 samples
  -> Added 49 new samples. Total: 3271
Train: 2780 | Val: 491
Class Weights: [7.998860836029053, 3.9367666244506836, 10.213482856750488, 4.5260114669799805, 11.450310707092285, 6.250929355621338]


Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at microsoft/deberta-v3-large and are newly initialized: ['classifier.bias', 'classifier.weight', 'pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step,Training Loss,Validation Loss,F1 Macro
100,1.295900,1.163968,0.327246
200,1.053900,0.949916,0.426026
300,0.864200,0.906712,0.467814
400,0.820000,0.943294,0.482197



⚖️  OPTIMIZING THRESHOLDS...


Label                | Best Thresh  | F1       | Prec     | Recall  
----------------------------------------------------------------------
stereotype           | 0.54         | 0.5150   | 0.3750   | 0.8219
vilification         | 0.54         | 0.6926   | 0.6012   | 0.8167
dehumanization       | 0.56         | 0.3655   | 0.2368   | 0.8000
extreme_language     | 0.53         | 0.6097   | 0.4881   | 0.8119
lack_of_empathy      | 0.35         | 0.3393   | 0.2147   | 0.8085
invalidation         | 0.59         | 0.4527   | 0.3216   | 0.7639

🔮 PREDICTING ON DEV SET...



 DONE! File saved: submission_deberta_large.csv


## Experiment Conclusion: Diminishing Returns

**Outcome:** **Did NOT beat the High Score (Score < 0.508)**

**Analysis:**
The Large model underperformed or matched the Base model, which is a common phenomenon in small datasets (~3k samples).
1.  **Overfitting Risk:** Large models have massive capacity. With only 3,000 training examples, `DeBERTa-Large` likely memorized the training data too quickly, failing to generalize to the subtle variations in the test set.
2.  **Batch Size Trade-off:** To fit the model on the GPU, we had to use a physical batch size of 4. Even with Gradient Accumulation, this can lead to noisy Batch Normalization statistics, making training less stable than the Base model (which used a healthy batch size of 16).
3.  **Weighting:** Removing the 1.5x aggressive weighting might have hurt Recall for rare classes. The Large model needed that "push" just as much as the Base model did.

**Final Verdict:**
For this specific dataset size, **DeBERTa-v3-Base** is the optimal architecture. It offers the perfect balance of semantic understanding and parameter efficiency. We will stick with the **Base model + Aggressive Weights** as our Champion.

# Experiment: Large Model Calibration (Final Fix)
**Model:** `microsoft/deberta-v3-large`
**Status:** **RECOVERY MODE**

**Objective:**
To fix the stability issues observed in the previous "Large Model" run (where Validation Loss spiked to 0.94).

**The Diagnosis:**
1.  **Overfitting:** The Large model memorized the training data too quickly. We are increasing `weight_decay` from `0.01` to `0.05` to force simpler decision boundaries.
2.  **Recall Collapse:** In the previous run, we removed the `1.5x` weight multiplier, which hurt Recall. We are **restoring the 1.5x Aggressive Weights** to force the model to pay attention to rare classes.
3.  **Threshold Limits:** The previous run stopped searching for thresholds at `0.60`. We are expanding the search range to `0.95` because large models tend to be over-confident (calibrated probabilities close to 1.0), so the optimal cutoff might be higher (e.g., 0.80).

**Hypothesis:** By controlling the overfitting (Regularization) while simultaneously pushing the gradients (Weights), we can make the Large model behave as stably as the Base model but with higher capacity.

In [ ]:
# =========================================================
#  SUBTASK 3: ENGLISH ONLY - RECALL BOOSTER (DEBERTA LARGE)
#  Environment: Google Colab
# =========================================================

from google.colab import drive
import os
import shutil
import pandas as pd
import torch
import torch.nn as nn
import numpy as np
import random
from torch.utils.data import Dataset
from transformers import (
    AutoTokenizer, AutoModelForSequenceClassification,
    Trainer, TrainingArguments, DataCollatorWithPadding,
    EarlyStoppingCallback
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score
import warnings

# 1. SETUP & INSTALLATION
# ---------------------------
print("⚙️ Installing libraries...")
# Using %%capture to hide the long output logs
!pip install transformers torch pandas scikit-learn tqdm sentencepiece accelerate -q

drive.mount('/content/drive')

# Set Seed for Reproducibility
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(42)
warnings.filterwarnings('ignore')

# 2. PATH CONFIGURATION & UNZIPPING
# ---------------------------
# We check if the folder exists. If not, we unzip it from Drive.
TARGET_DIR = 'subtask3' # The folder inside the zip
ZIP_PATH = '/content/drive/My Drive/data/NLP_AIMS/dev_phase.zip'

if not os.path.exists(TARGET_DIR):
    print(f"📂 Unzipping dataset from: {ZIP_PATH}")
    # We unzip into the current directory
    !unzip -o -q "{ZIP_PATH}"
else:
    print("✅ Dataset already extracted.")

# IMPORTANT: Based on your previous code, we assume the unzipped folder
# acts as the base. If the zip extracts into current dir, BASE_PATH is '.'
# If it extracts a folder named 'subtask3', we point there.
if os.path.exists('subtask3/train/eng.csv'):
    BASE_PATH = 'subtask3'
elif os.path.exists('train/eng.csv'):
    BASE_PATH = '.'
else:
    # Fallback search if folder structure is different
    # (Sometimes zips have extra nested folders)
    print("⚠️  Searching for 'eng.csv'...")
    for root, dirs, files in os.walk("."):
        if "eng.csv" in files and "train" in root:
            BASE_PATH = os.path.dirname(os.path.dirname(root))
            print(f"Found data at: {BASE_PATH}")
            break

print(f"📍 BASE_PATH set to: {BASE_PATH}")

# 3. MODEL CONFIGURATION (DEBERTA LARGE OPTIMIZED)
# ---------------------------
CONFIG = {
    'model': 'microsoft/deberta-v3-large',  # <--- LARGE MODEL
    'max_len': 256,
    'epochs': 5,
    'lr': 7e-6,            # <--- Low LR for stability
    'batch': 4,            # <--- Small Batch for Memory
    'grad_accum': 8,       # <--- Accumulate to simulate batch=32
    'warmup': 0.1,
    'weight_decay': 0.01,
    'augment_factor': 4,
    'label_smoothing': 0.05,
}

LABELS = ['stereotype', 'vilification', 'dehumanization', 'extreme_language', 'lack_of_empathy', 'invalidation']

# 4. SYNONYMS & AUGMENTATION
# ---------------------------
SYNONYMS = {
    'stereotype': {'typical': ['usual', 'classic'], 'always': ['constantly', 'forever'], 'all': ['every', 'entire'],
                   'lazy': ['useless', 'idle'], 'stupid': ['dumb', 'idiotic'], 'criminal': ['thug', 'gangster']},
    'vilification': {'evil': ['vile', 'wicked'], 'scum': ['trash', 'garbage'], 'vermin': ['rats', 'pests'],
                     'monster': ['demon', 'devil'], 'worthless': ['pathetic', 'disgusting']},
    'dehumanization': {'animal': ['dog', 'beast'], 'subhuman': ['savage', 'primitive'],
                       'infestation': ['plague', 'disease'], 'breed': ['spawn', 'multiply']},
    'extreme_language': {'destroy': ['crush', 'wipe out'], 'exterminate': ['kill', 'eradicate'],
                         'massacre': ['genocide', 'bloodbath'], 'purge': ['cleanse', 'remove']},
    'lack_of_empathy': {'deserve': ['asked for it', 'karma'], 'fault': ['blame', 'own doing'],
                        'who cares': ['irrelevant', 'whatever'], 'deal with it': ['suck it up', 'cry more']},
    'invalidation': {'exaggerate': ['drama queen', 'overreacting'], 'fake': ['hoax', 'staged'],
                     'lie': ['bullshit', 'myth'], 'pretending': ['acting', 'playing victim']}
}

def augment_text(text, category, prob=0.35):
    cat_syns = SYNONYMS.get(category, {})
    if not cat_syns: return text
    words = text.split()
    aug_words = []
    for word in words:
        w_lower = word.lower().strip('.,!?;:')
        if w_lower in cat_syns and random.random() < prob:
            syn = random.choice(cat_syns[w_lower])
            if word[0].isupper(): syn = syn.capitalize()
            aug_words.append(syn)
        else:
            aug_words.append(word)
    return ' '.join(aug_words)

def augment_df(df, factor=3):
    print(f"Augmenting English Data: {len(df)} samples")
    aug_texts, aug_labels = [], []
    for _, row in df.iterrows():
        text = row['text']
        label_vals = {label: row[label] for label in LABELS}
        pos_labels = [l for l, v in label_vals.items() if v == 1]

        if pos_labels:
            for _ in range(factor):
                cat = random.choice(pos_labels)
                aug_text = augment_text(text, cat)
                if aug_text != text:
                    aug_texts.append(aug_text)
                    aug_labels.append(label_vals.copy())

    if aug_texts:
        aug_df = pd.DataFrame({'text': aug_texts, **{label: [d[label] for d in aug_labels] for label in LABELS}})
        result = pd.concat([df, aug_df], ignore_index=True)
        print(f"  -> Added {len(aug_df)} new samples. Total: {len(result)}")
        return result
    return df

# 5. CLASSES: DATASET & TRAINER
# ---------------------------
class WeightedLossTrainer(Trainer):
    def __init__(self, class_weights, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_weights = class_weights

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        logits = outputs.logits
        # Move weights to same device as logits
        weights = self.class_weights.to(logits.device)
        loss_fct = nn.BCEWithLogitsLoss(pos_weight=weights)
        loss = loss_fct(logits, labels)
        return (loss, outputs) if return_outputs else loss

class PolarizationDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts; self.labels = labels; self.tokenizer = tokenizer; self.max_len = max_len
    def __len__(self): return len(self.texts)
    def __getitem__(self, idx):
        enc = self.tokenizer(str(self.texts[idx]), truncation=True, max_length=self.max_len, padding="max_length", return_tensors='pt')
        item = {k: v.squeeze(0) for k, v in enc.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.float)
        return item

def compute_metrics(pred):
    probs = torch.sigmoid(torch.from_numpy(pred.predictions)).numpy()
    preds = (probs > 0.5).astype(int)
    return {'f1_macro': f1_score(pred.label_ids, preds, average='macro', zero_division=0)}

# 6. MAIN PIPELINE
# ---------------------------
# =========================================================
#  FINAL FIX: UNLOCKED THRESHOLDS + HIGH RECALL WEIGHTS
# =========================================================

def train_english_optimized():
    print("="*60)
    print("🚀 STARTING DEBERTA-V3-LARGE (FINAL FIX)")
    print("="*60)

    # A. Load & Augment
    train_file = os.path.join(BASE_PATH, "train/eng.csv")
    if not os.path.exists(train_file):
        raise FileNotFoundError(f"Could not find training file at: {train_file}")

    df_raw = pd.read_csv(train_file)
    df_aug = augment_df(df_raw, factor=CONFIG['augment_factor'])

    # B. Split
    train_df, val_df = train_test_split(df_aug, test_size=0.15, random_state=42)
    print(f"Train: {len(train_df)} | Val: {len(val_df)}")

    # C. Calculate Class Weights (RESTORED TO 1.5)
    # We need this aggression to keep Recall high (0.74)
    pos_counts = train_df[LABELS].sum()
    neg_counts = len(train_df) - pos_counts
    weights = (neg_counts / (pos_counts + 1e-6)) * 1.5
    class_weights = torch.tensor(weights.values, dtype=torch.float)
    print(f"Class Weights: {class_weights.tolist()}")

    # D. Model & Tokenizer
    tokenizer = AutoTokenizer.from_pretrained(CONFIG['model'])
    model = AutoModelForSequenceClassification.from_pretrained(CONFIG['model'], num_labels=6, problem_type="multi_label_classification")

    model.config.use_cache = False # Required for gradient checkpointing

    train_ds = PolarizationDataset(train_df['text'].tolist(), train_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])
    val_ds = PolarizationDataset(val_df['text'].tolist(), val_df[LABELS].values.tolist(), tokenizer, CONFIG['max_len'])

    args = TrainingArguments(
        output_dir="./results_deberta_large_final",
        num_train_epochs=CONFIG['epochs'],
        learning_rate=CONFIG['lr'],
        per_device_train_batch_size=CONFIG['batch'],
        gradient_accumulation_steps=CONFIG['grad_accum'],
        gradient_checkpointing=True,
        gradient_checkpointing_kwargs={'use_reentrant': False},
        fp16=True,
        warmup_ratio=CONFIG['warmup'],
        weight_decay=0.05, # Increased slightly to reduce overfitting (Val loss 0.94)
        eval_strategy="steps", eval_steps=100,
        save_strategy="steps", save_steps=100,
        load_best_model_at_end=True, metric_for_best_model="f1_macro",
        report_to="none", save_total_limit=1,
        logging_steps=50
    )

    trainer = WeightedLossTrainer(
        class_weights=class_weights,
        model=model, args=args,
        train_dataset=train_ds, eval_dataset=val_ds,
        compute_metrics=compute_metrics,
        data_collator=DataCollatorWithPadding(tokenizer),
        callbacks=[EarlyStoppingCallback(early_stopping_patience=4)]
    )

    trainer.train()

    # E. THRESHOLD OPTIMIZATION (THE FIX)
    print("\n⚖️  OPTIMIZING THRESHOLDS (Range Expanded: 0.15 - 0.95)...")
    val_preds = trainer.predict(val_ds)
    val_probs = torch.sigmoid(torch.tensor(val_preds.predictions)).numpy()

    best_thresholds = []
    print(f"{'Label':<20} | {'Best Thresh':<12} | {'F1':<8} | {'Prec':<8} | {'Recall':<8}")
    print("-" * 70)

    for i, label in enumerate(LABELS):
        best_t, best_f1, best_p, best_r = 0.5, 0, 0, 0

        # 🔴 CHANGED: Search up to 0.95 instead of stopping at 0.60
        for t in np.arange(0.15, 0.95, 0.01):
            preds = (val_probs[:, i] > t).astype(int)
            f1 = f1_score(val_df[LABELS].values[:, i], preds, zero_division=0)

            # Prefer higher threshold if F1 is equal (boosts precision)
            if f1 > best_f1:
                best_f1 = f1
                best_t = t
                best_p = precision_score(val_df[LABELS].values[:, i], preds, zero_division=0)
                best_r = recall_score(val_df[LABELS].values[:, i], preds, zero_division=0)

        best_thresholds.append(best_t)
        print(f"{label:<20} | {best_t:.2f}         | {best_f1:.4f}   | {best_p:.4f}   | {best_r:.4f}")

    # F. Predict on Dev
    print("\n🔮 PREDICTING ON DEV SET...")
    dev_file = os.path.join(BASE_PATH, "dev/eng.csv")
    if not os.path.exists(dev_file):
         raise FileNotFoundError(f"Could not find dev file at: {dev_file}")

    dev_df = pd.read_csv(dev_file)
    dev_ds = PolarizationDataset(dev_df['text'].tolist(), [[0]*6]*len(dev_df), tokenizer, CONFIG['max_len'])

    dev_preds = trainer.predict(dev_ds)
    dev_probs = torch.sigmoid(torch.tensor(dev_preds.predictions)).numpy()

    final_preds = (dev_probs > np.array(best_thresholds)).astype(int)

    sub_df = pd.DataFrame(final_preds, columns=LABELS)
    sub_df.insert(0, 'id', dev_df['id'])
    sub_df.to_csv("submission_deberta_large_fixed.csv", index=False)

    print("\n DONE! File saved: submission_deberta_large_fixed.csv")

if __name__ == "__main__":
    train_english_optimized()

⚙️ Installing libraries...
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✅ Dataset already extracted.
📍 BASE_PATH set to: subtask3
🚀 STARTING DEBERTA-V3-LARGE TRAINING (FIXED)
Augmenting English Data: 3222 samples
  -> Added 49 new samples. Total: 3271
Train: 2780 | Val: 491
Class Weights: [5.332573890686035, 2.6245110034942627, 6.808988571166992, 3.017341136932373, 7.633540153503418, 4.167286396026611]


Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at microsoft/deberta-v3-large and are newly initialized: ['classifier.bias', 'classifier.weight', 'pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step,Training Loss,Validation Loss,F1 Macro
100,1.042000,0.848370,0.453123
200,0.863800,0.735016,0.486122
300,0.711200,0.743074,0.474338
400,0.675800,0.783008,0.506388



⚖️  OPTIMIZING THRESHOLDS...


Label                | Best Thresh  | F1       | Prec     | Recall  
----------------------------------------------------------------------
stereotype           | 0.52         | 0.5405   | 0.4027   | 0.8219
vilification         | 0.59         | 0.7239   | 0.6554   | 0.8083
dehumanization       | 0.39         | 0.3663   | 0.2357   | 0.8222
extreme_language     | 0.58         | 0.6215   | 0.5200   | 0.7723
lack_of_empathy      | 0.53         | 0.3500   | 0.2288   | 0.7447
invalidation         | 0.59         | 0.4753   | 0.3510   | 0.7361

🔮 PREDICTING ON DEV SET...



 DONE! File saved: submission_deberta_large.csv


## Experiment Conclusion: Capacity vs. Data Size

**Outcome:** **Did NOT beat the Base Model High Score (0.508)**

**Final Analysis:**
Despite the fixes (High Regularization, Aggressive Weights, Expanded Thresholds), the Large model essentially matched the Base model but took 3x longer to train.
* **The "Data Wall":** With only ~3,000 samples, we have likely hit the "Data Wall." The dataset is simply too small to justify the 304M parameters of `DeBERTa-Large`. The extra parameters are just fitting noise rather than signal.
* **Optimization Difficulty:** Large models are notoriously hard to fine-tune on small batches (Batch Size 4). The noise in the gradients likely prevented the model from finding the global optimum that the Base model found easily with Batch Size 16.

**Strategic Decision:**
We formally retire the `Large` architecture for this task. The **Champion Model remains `DeBERTa-v3-Base`** (from the previous experiment) due to its efficiency and superior generalization on this specific dataset size.